File: scripts/lines_report.py
#!/usr/bin/env python3
"""
Gera relatório de contagem de linhas por arquivo e por módulo (top-level folder).
Ignora arquivos listados pelo Git (.gitignore) se o diretório for um repositório Git.

Uso:
  python scripts/lines_report.py

Saída: impressa no stdout; redirecione para arquivo se desejar.
"""
import os
import subprocess
import sys
from collections import defaultdict
from datetime import datetime


def repo_root():
    # script está em scripts/; repo root é um nível acima
    return os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))


def git_list_files(root):
    try:
        # verifica se estamos dentro de um work tree
        p = subprocess.run(['git', 'rev-parse', '--is-inside-work-tree'], cwd=root, capture_output=True, text=True)
        if p.returncode != 0 or p.stdout.strip() != 'true':
            return None
        p2 = subprocess.run(['git', 'ls-files', '--cached', '--others', '--exclude-standard'], cwd=root, capture_output=True, text=True)
        if p2.returncode != 0:
            return None
        files = [line.strip() for line in p2.stdout.splitlines() if line.strip()]
        return files
    except FileNotFoundError:
        return None


def parse_gitignore(root):
    path = os.path.join(root, '.gitignore')
    patterns = []
    if not os.path.exists(path):
        return patterns
    with open(path, 'r', encoding='utf-8', errors='ignore') as f:
        for ln in f:
            ln = ln.strip()
            if not ln or ln.startswith('#'):
                continue
            patterns.append(ln)
    return patterns


def match_simple_patterns(relpath, patterns):
    # fallback simples: usa fnmatch para casamentos básicos e prefixo para diretórios
    import fnmatch
    for p in patterns:
        if p.endswith('/'):
            if relpath.startswith(p.rstrip('/')):
                return True
        # tratar padrão com /** ou /* simplificado
        if fnmatch.fnmatch(relpath, p) or fnmatch.fnmatch(relpath, os.path.join(p, '*')):
            return True
    return False


def fallback_walk(root):
    patterns = parse_gitignore(root)
    files = []
    for dirpath, dirnames, filenames in os.walk(root):
        # pular .git
        rel_dir = os.path.relpath(dirpath, root)
        if rel_dir == '.':
            rel_dir = ''
        if rel_dir.split(os.sep)[0] == '.git':
            continue
        for fn in filenames:
            rel = os.path.normpath(os.path.join(rel_dir, fn)).lstrip('.' + os.sep)
            if rel.startswith('.git'):
                continue
            if patterns and match_simple_patterns(rel, patterns):
                continue
            files.append(rel)
    return files


def count_lines(fullpath):
    try:
        with open(fullpath, 'rb') as f:
            # contar linhas lendo por iterador (funciona mesmo em binários)
            return sum(1 for _ in f)
    except Exception:
        return 0


def main():
    root = repo_root()
    files = git_list_files(root)
    if files is None:
        files = fallback_walk(root)

    # filtrar arquivos: garantir que existam, e excluir tudo dentro de Backup/ e imagens .jpg/.jpeg
    filtered_files = []
    for f in files:
        fullp = os.path.join(root, f)
        if not os.path.isfile(fullp):
            continue
        parts = f.split(os.sep)
        if parts and parts[0].lower() == 'backup':
            continue
        low = f.lower()
        if low.endswith('.jpg') or low.endswith('.jpeg'):
            continue
        filtered_files.append(f)

    module_stats = defaultdict(lambda: {'lines': 0, 'files': 0, 'file_list': []})
    total_lines = 0
    total_files = 0

    for rel in filtered_files:
        full = os.path.join(root, rel)
        lines = count_lines(full)
        total_lines += lines
        total_files += 1
        parts = rel.split(os.sep)
        module = parts[0] if len(parts) > 1 else (parts[0] if parts[0] else 'root')
        module_stats[module]['lines'] += lines
        module_stats[module]['files'] += 1
        module_stats[module]['file_list'].append((rel, lines))

    # ranking por linhas
    ranking = sorted(module_stats.items(), key=lambda kv: kv[1]['lines'], reverse=True)

    out_lines = []
    out_lines.append('Relatório de linhas por arquivo e por módulo')
    out_lines.append(f'Gerado em: {datetime.now().isoformat()}')
    out_lines.append(f'Raiz do repositório: {root}')
    out_lines.append(f'Arquivos considerados (ignora .gitignore quando há Git): {len(filtered_files)}')
    out_lines.append(f'Total de arquivos: {total_files}  Total de linhas: {total_lines}')
    out_lines.append('')
    out_lines.append('Ranking de módulos por linhas:')
    for i, (mod, stats) in enumerate(ranking, 1):
        out_lines.append(f"{i:2d}. {mod:30s} — {stats['lines']:8d} linhas  / {stats['files']:5d} arquivos")

    out_lines.append('')
    out_lines.append('Top 10 arquivos por linhas (global):')
    all_files = []
    for mod, s in module_stats.items():
        all_files.extend(s['file_list'])
    all_files.sort(key=lambda t: t[1], reverse=True)
    for rel, ln in all_files[:10]:
        out_lines.append(f"{ln:8d}  {rel}")

    # grava em arquivo na raiz do repositório
    outpath = os.path.join(root, 'lines_report.txt')
    try:
        with open(outpath, 'w', encoding='utf-8') as f:
            f.write('\n'.join(out_lines))
        print(f'Relatório gravado em: {outpath}')
    except Exception as e:
        print('Erro ao gravar relatório:', e)


if __name__ == '__main__':
    main()

File: scripts/lines_report.php
<?php
/**
 * Gera relatório de contagem de linhas por arquivo e por módulo (top-level folder).
 * Ignora arquivos listados pelo Git (.gitignore) se o diretório for um repositório Git.
 * Uso:
 *   php scripts/lines_report.php
 * Saída: impresso no stdout; redirecione para arquivo se desejar.
 */

function repo_root()
{
    return realpath(__DIR__ . DIRECTORY_SEPARATOR . '..');
}

function git_list_files($root)
{
    $cwd = $root;
    $out = [];
    $rc = null;
    exec('git -C ' . escapeshellarg($cwd) . ' rev-parse --is-inside-work-tree 2>&1', $out, $rc);
    if ($rc !== 0 || trim(implode("\n", $out)) !== 'true') {
        return null;
    }
    $files = [];
    exec('git -C ' . escapeshellarg($cwd) . ' ls-files --cached --others --exclude-standard 2>&1', $files, $rc);
    if ($rc !== 0) {
        return null;
    }
    return $files;
}

function parse_gitignore($root)
{
    $p = $root . DIRECTORY_SEPARATOR . '.gitignore';
    $patterns = [];
    if (!file_exists($p)) return $patterns;
    $lines = file($p, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
    foreach ($lines as $ln) {
        $ln = trim($ln);
        if ($ln === '' || strpos($ln, '#') === 0) continue;
        $patterns[] = $ln;
    }
    return $patterns;
}

function match_simple_patterns($rel, $patterns)
{
    foreach ($patterns as $p) {
        if (substr($p, -1) === '/') {
            if (strpos($rel, rtrim($p, '/')) === 0) return true;
        }
        if (fnmatch($p, $rel) || fnmatch($p . '/*', $rel)) return true;
    }
    return false;
}

function fallback_walk($root)
{
    $patterns = parse_gitignore($root);
    $files = [];
    $it = new RecursiveIteratorIterator(new RecursiveDirectoryIterator($root, FilesystemIterator::SKIP_DOTS));
    foreach ($it as $fileinfo) {
        $path = $fileinfo->getPathname();
        $rel = ltrim(str_replace($root, '', $path), DIRECTORY_SEPARATOR);
        if (strpos($rel, '.git' . DIRECTORY_SEPARATOR) === 0) continue;
        if (!empty($patterns) && match_simple_patterns($rel, $patterns)) continue;
        $files[] = $rel;
    }
    return $files;
}

function count_lines($fullpath)
{
    $cnt = 0;
    $fp = @fopen($fullpath, 'rb');
    if (!$fp) return 0;
    while (!feof($fp)) {
        fgets($fp);
        $cnt++;
    }
    fclose($fp);
    return $cnt;
}

$root = repo_root();
$files = git_list_files($root);
if ($files === null) {
    $files = fallback_walk($root);
}

$filtered = [];
foreach ($files as $f) {
    $full = $root . DIRECTORY_SEPARATOR . $f;
    if (!is_file($full)) continue;
    $parts = explode(DIRECTORY_SEPARATOR, $f);
    if (isset($parts[0]) && strtolower($parts[0]) === 'backup') continue;
    $ext = strtolower(pathinfo($f, PATHINFO_EXTENSION));
    if ($ext === 'jpg' || $ext === 'jpeg') continue;
    $filtered[] = $f;
}

$module_stats = [];
$total_files = 0;
$total_lines = 0;

foreach ($filtered as $rel) {
    $full = $root . DIRECTORY_SEPARATOR . $rel;
    $lines = count_lines($full);
    $total_files++;
    $total_lines += $lines;
    $parts = explode(DIRECTORY_SEPARATOR, $rel);
    $module = count($parts) > 1 ? $parts[0] : ($parts[0] !== '' ? $parts[0] : 'root');
    if (!isset($module_stats[$module])) $module_stats[$module] = ['lines' => 0, 'files' => 0, 'list' => []];
    $module_stats[$module]['lines'] += $lines;
    $module_stats[$module]['files'] += 1;
    $module_stats[$module]['list'][] = [$rel, $lines];
}

// sort modules by lines desc
uasort($module_stats, function($a, $b) { return $b['lines'] - $a['lines']; });

echo "Relatório de linhas por arquivo e por módulo\n";
echo "Raiz do repositório: $root\n";
echo "Arquivos considerados (ignora .gitignore quando há Git): " . count($filtered) . "\n";
echo "Total de arquivos: $total_files  Total de linhas: $total_lines\n\n";

echo "Ranking de módulos por linhas:\n";
$i = 1;
foreach ($module_stats as $mod => $s) {
    printf("%2d. %-30s — %8d linhas  / %5d arquivos\n", $i, $mod, $s['lines'], $s['files']);
    $i++;
}

echo "\nTop 10 arquivos por linhas (global):\n";
$all = [];
foreach ($module_stats as $s) {
    foreach ($s['list'] as $t) $all[] = $t;
}
usort($all, function($a, $b) { return $b[1] - $a[1]; });
for ($k = 0; $k < min(10, count($all)); $k++) {
    printf("%8d  %s\n", $all[$k][1], $all[$k][0]);
}
